ext/opentelemetry-ext-jinja2/src/opentelemetry/ext/jinja2/__init__.py

```diff
@@ -0,0 +1,128 @@
+    from opentelemetry.ext.jinja2 import Jinja2Instrumentor
+
+    trace.set_tracer_provider(TracerProvider())
+
+    Jinja2Instrumentor().instrument()
+
+    env = Environment(loader=FileSystemLoader("templates"))
+    template = env.get_template('mytemplate.html')
+
+API
+---
+"""
+
+import logging
+
+import jinja2
+from wrapt import ObjectProxy
+from wrapt import wrap_function_wrapper as _wrap
+
+from opentelemetry import trace
+from opentelemetry.auto_instrumentation.instrumentor import BaseInstrumentor
+from opentelemetry.ext.jinja2.version import __version__
+from opentelemetry.trace.status import Status, StatusCanonicalCode
+
+logger = logging.getLogger(__name__)
+
+ATTRIBUTE_JINJA2_TEMPLATE_NAME = "jinja2.template_name"
+ATTRIBUTE_JINJA2_TEMPLATE_PATH = "jinja2.template_path"
+DEFAULT_TEMPLATE_NAME = "<memory>"
+
+
+def _wrap_render(wrapped, instance, args, kwargs):
+    """Wrap `Template.render()` or `Template.generate()`
+    """
+    tracer = trace.get_tracer(__name__, __version__)
+    template_name = instance.name or DEFAULT_TEMPLATE_NAME
+    attributes = {ATTRIBUTE_JINJA2_TEMPLATE_NAME: template_name}
+    with tracer.start_as_current_span(
+        "jinja2.render", kind=trace.SpanKind.INTERNAL, attributes=attributes
+    ):
+        return wrapped(*args, **kwargs)
+
+
+def _wrap_compile(wrapped, _, args, kwargs):
+    tracer = trace.get_tracer(__name__, __version__)
+    template_name = (
+        args[1] if len(args) > 1 else kwargs.get("name", DEFAULT_TEMPLATE_NAME)
+    )
+    attributes = {ATTRIBUTE_JINJA2_TEMPLATE_NAME: template_name}
+    with tracer.start_as_current_span(
+        "jinja2.compile", kind=trace.SpanKind.INTERNAL, attributes=attributes
+    ):
+        return wrapped(*args, **kwargs)
+
+
+def _wrap_load_template(wrapped, _, args, kwargs):
+    tracer = trace.get_tracer(__name__, __version__)
+    template_name = kwargs.get("name", args[0])
+    attributes = {ATTRIBUTE_JINJA2_TEMPLATE_NAME: template_name}
+    with tracer.start_as_current_span(
+        "jinja2.load", kind=trace.SpanKind.INTERNAL, attributes=attributes
+    ) as span:
+        template = None
+        try:
+            template = wrapped(*args, **kwargs)
+            return template
+        finally:
+            if template:
+                span.set_attribute(
+                    ATTRIBUTE_JINJA2_TEMPLATE_PATH, template.filename
+                )
+
+
+def _unwrap(obj, attr):
+    func = getattr(obj, attr, None)
+    if func and isinstance(func, ObjectProxy) and hasattr(func, "__wrapped__"):
+        setattr(obj, attr, func.__wrapped__)
+
+
+class Jinja2Instrumentor(BaseInstrumentor):
+    """A instrumentor for jinja2
+
+    See `BaseInstrumentor`
+    """
+
+    def _instrument(self, **kwargs):
+        _wrap(jinja2, "environment.Template.render", _wrap_render)
+        _wrap(jinja2, "environment.Template.generate", _wrap_render)
+        _wrap(jinja2, "environment.Environment.compile", _wrap_compile)
+        _wrap(
```

Comment:
We are trying to follow an approach where a `TracerProvider` can be passed to `_instrument`, then that tracer provider is used in `trace.get_tracer()` to the the specific tracer.
You can get some inspiration from https://github.com/open-telemetry/opentelemetry-python/blob/b5b297c7606fdf851c32f413968d842aa7e23b60/ext/opentelemetry-ext-pymysql/src/opentelemetry/ext/pymysql/__init__.py#L56.